src/background-monitor.c

```diff
@@ -0,0 +1,505 @@
+{
+  DoneData *ddata;
+  g_autofree char *name = flatpak_instance_get_display_name (instance);
+  g_autofree char *handle = NULL;
+  static int count;
+
+  ddata = g_new (DoneData, 1);
+  ddata->app_id = g_strdup (flatpak_instance_get_app (instance));
+  ddata->perm = permission;
+
+  g_debug ("Notify background for %s", ddata->app_id);
+
+  handle = g_strdup_printf ("/org/freedesktop/portal/desktop/notify/background%d", count++);
+
+  add_background_app (ddata->app_id, handle);
+
+  xdp_impl_background_call_notify_background (background,
+                                              handle,
+                                              ddata->app_id,
+                                              name,
+                                              NULL,
+                                              notify_background_done,
+                                              ddata);
+}
+
+static void
+check_background_apps (GTask *task,
+                       gpointer source_object,
+                       gpointer task_data,
+                       GCancellable *cancellable)
+{
+  g_autoptr(GVariant) perms = NULL;
+  g_autoptr(GHashTable) app_states = NULL;
+  g_autoptr(GPtrArray) apps = NULL;
+  int i;
+
+  app_states = get_app_states ();
+
+  if (app_states == NULL)
+    return;
+
+  g_debug ("Checking background permissions");
+  perms = get_permissions ();
+  apps = flatpak_instance_get_all ();
+
+  remove_outdated_applications (apps);
+
+  for (i = 0; i < apps->len; i++)
+    {
+      FlatpakInstance *instance = g_ptr_array_index (apps, i);
+      const char *app_id = flatpak_instance_get_app (instance);
+      Permission permission;
+      AppState state;
+      const char *state_names[] = { "background", "running", "active" };
+      g_autofree char *handle = NULL;
+
+      if (!flatpak_instance_is_running (instance))
+        continue;
+
+      state = get_app_state (app_id, app_states);
+      g_debug ("App %s is %s", app_id, state_names[state]);
+
+      if (state != BACKGROUND)
+        {
+          remove_background_app (app_id);
+          continue;
+        }
+
+      if (!lookup_background_app (app_id, &handle))
+        {
+          add_background_app (app_id, NULL);
+          continue;
+        }
+
+      if (handle)
+        continue; /* already notified */
+
+      permission = get_permission (app_id, perms);
+      if (permission == NO)
+        {
+          pid_t pid = flatpak_instance_get_child_pid (instance);
+          g_debug ("Killing app %s (child pid %u)", app_id, pid);
+          kill (pid, SIGKILL);
+        }
+      else if (permission == ASK || permission == UNSET)
+        {
+          send_notification (instance, permission);
+        }
+    }
+}
+
+static gboolean
+enforce_background_permissions (gpointer data)
+{
+  g_autoptr(GTask) task = NULL;
+
+  task = g_task_new (NULL, NULL, NULL, NULL);
+  g_task_run_in_thread (task, check_background_apps);
+
+  return G_SOURCE_CONTINUE;
```

Comment:
This seems risky to me. You run this in a regular 30sec timeout and then the timeout spawns a thread. However, the thread relies on global state (the applications hashtable). So, you risk some weird behaviour if ever check_background_apps takes > 30 secs for whatever reason. I think it would be better if enforce_background_permissions returned FALSE, and the thread instead queued a new timeout.

Reply:
> This seems risky to me. You run this in a regular 30sec timeout and then the timeout spawns a thread. However, the thread relies on global state (the applications hashtable). So, you risk some weird behaviour if ever check_background_apps takes > 30 secs for whatever reason. I think it would be better if enforce_background_permissions returned FALSE, and the thread instead queued a new timeout.

I've changed this - no need for a timeout at all